File: AutoPecas.Web/ClientApp/src/app/service/usuario.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { FiltroSpec } from '../model/geral/filtro-spec.model';
import { PaginacaoResultado } from '../model/geral/paginacao-resultado.model';
import Usuario from '../model/usuario/usuario.model';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root',
})

export class UsuarioService {
  private readonly apiEndpoint = `${environment.apiUrl}usuario`;

  constructor(private httpClient: HttpClient) { }

  public listar(filtro: FiltroSpec) {
    return this.httpClient.post<PaginacaoResultado<Usuario>>(`${this.apiEndpoint}/lista/`, filtro);
  }

  public busca(texto: string): Observable<Usuario[]> {
    return this.httpClient.get<Usuario[]>(`${this.apiEndpoint}/busca/${texto}`);
  }

  public obter(id: number): Observable<Usuario> {
    return this.httpClient.get<Usuario>(`${this.apiEndpoint}/${id}`);
  }

  public incluir(usuario: Usuario): Observable<number> {
    console.log(usuario);

    return this.httpClient.post<number>(`${this.apiEndpoint}`, usuario);
  }

  public editar(usuario: Usuario): Observable<number> {
    console.log(usuario);

    return this.httpClient.put<number>(`${this.apiEndpoint}`, usuario);
  }

  public login(usuario: Usuario): Observable<Usuario> {
    console.log(usuario);
    
    return this.httpClient.post<Usuario>(`${this.apiEndpoint}/login`, usuario);
  }
}

File: AutoPecas.Service/UsuarioService.cs
using Autopecas.Infra.Data;
using AutoPecas.Core.Model;
using AutoPecas.Core.Spec;
using LinqKit;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace AutoPecas.Service
{
    public class UsuarioService
    {
        private readonly AutoPecasDbContext _AutoPecasDbContext;

        private void AplicarFiltro(FiltroSpec filtro, out IQueryable<Usuario> query)
        {
            query = _AutoPecasDbContext.Usuarios
                        .AsExpandableEFCore();

            var predicate = PredicateBuilder.New<Usuario>(true);

            if (filtro.Filtros.TryGetValue("nome", out var nome))
            {
                predicate.And(p => p.Nome.Contains((string)nome));
            }

            if (filtro.Filtros.TryGetValue("usuario", out var usuario))
            {
                predicate.Or(p => p.NomeUsuario.Contains((string)usuario));
            }

            query = query.Where(predicate);
        }

        public async Task<Usuario> Login(Usuario usuario)
        {
            return await _AutoPecasDbContext
                .Usuarios
                .Where(p => p.NomeUsuario == usuario.NomeUsuario && p.Senha == usuario.Senha && p.Ativo == true)
                .FirstOrDefaultAsync();
        }

        public UsuarioService(AutoPecasDbContext autoPecasDbContext)
        {
            _AutoPecasDbContext = autoPecasDbContext;
        }

        public async Task<PaginacaoResultado<Usuario>> Lista(FiltroSpec filtro)
        {
            AplicarFiltro(filtro, out var query);

            var resultado = new PaginacaoResultado<Usuario>(query, filtro.Pagina, filtro.Tamanho, filtro.Total);

            resultado.Lista = await query
                .Skip(filtro.Tamanho * (filtro.Pagina - 1))
                .Take(filtro.Tamanho)
                .ToListAsync();

            return resultado;
        }

        public async Task<Usuario> Obter(int idUsuario)
        {
            return await _AutoPecasDbContext
                .Usuarios
                .Where(p => p.Id == idUsuario)
                .FirstOrDefaultAsync();
        }

        public async Task<int> Incluir(Usuario usuario)
        {
            _AutoPecasDbContext.Add(usuario);

            return await _AutoPecasDbContext.SaveChangesAsync();
        }

        public async Task<int> Editar(Usuario usuario)
        {
            _AutoPecasDbContext.Update(usuario);
            return await _AutoPecasDbContext.SaveChangesAsync();
        }

        public async Task<IList<Usuario>> Busca(string texto)
        {
            return await _AutoPecasDbContext
                .Usuarios
                .Where(p => p.Nome.Contains(texto))
                .Take(5)
                .ToListAsync();
        }
    }
}

